pkg/driver/node/mounter/pod_mounter.go

```diff
@@ -0,0 +1,403 @@
+	podID, volumeID := credentialCtx.PodID, credentialCtx.VolumeID
+
+	err := pm.checkTargetPath(target)
+	if err != nil {
+		return fmt.Errorf("Failed to check target path %q: %w", target, err)
+	}
+
+	isMountPoint, err := pm.IsMountPoint(target)
+	if err != nil {
+		return fmt.Errorf("Could not check if %q is a mount point: %w", target, err)
+	}
+
+	// TODO: If `target` is a `systemd`-mounted Mountpoint, this would return an error,
+	// but we should still update the credentials for it by calling `credProvider.Provide`.
+	pod, podPath, err := pm.waitForMountpointPod(ctx, podID, volumeID)
+	if err != nil {
+		return fmt.Errorf("Failed to wait for Mountpoint Pod for %q: %w", target, err)
+	}
+
+	// TODO: After this point we know Mountpoint Pod's name, we should add some debugging tips to our error messages returned, like:
+	// """
+	// Failed to mount on "..." because "..."`
+	// You can see logs for Mountpoint Pod by running:
+	// `kubectl logs -n mount-s3 "mp-..." --previous`
+	// """
+
+	podCredentialsPath, err := pm.ensureCredentialsDirExists(podPath)
+	if err != nil {
+		return fmt.Errorf("Failed to create credentials directory for %q: %W", target, err)
+	}
+
+	credentialCtx.SetWriteAndEnvPath(podCredentialsPath, mppod.PathInsideMountpointPod(mppod.KnownPathCredentials))
+
+	// Note that this part happens before `isMountPoint` check, as we want to update credentials even though
+	// there is an existing mount point at `target`.
+	credEnv, authenticationSource, err := pm.credProvider.Provide(ctx, credentialCtx)
+	if err != nil {
+		klog.V(4).Infof("NodePublishVolume: Failed to provide credentials for %s: %v", target, err)
+		return fmt.Errorf("Failed to provide credentials for %q: %w", target, err)
+	}
+
+	if isMountPoint {
+		klog.V(4).Infof("Target path %q is already mounted", target)
+		return nil
+	}
+
+	env := envprovider.Default()
+	env.Merge(credEnv)
+
+	// Move `--aws-max-attempts` to env if provided
+	if maxAttempts, ok := args.Remove(mountpoint.ArgAWSMaxAttempts); ok {
+		env.Set(envprovider.EnvMaxAttempts, maxAttempts)
+	}
+
+	args.Set(mountpoint.ArgUserAgentPrefix, UserAgent(authenticationSource, pm.kubernetesVersion))
+
+	podMountSockPath := mppod.PathOnHost(podPath, mppod.KnownPathMountSock)
+	podMountErrorPath := mppod.PathOnHost(podPath, mppod.KnownPathMountError)
+
+	klog.V(4).Infof("Waiting for Mountpoint Pod %s to be ready to accept connections on %s", pod.Name, podMountSockPath)
+
+	// Wait until Mountpoint Pod is ready to accept connections
+	err = util.WaitForUnixSocket(mountpointPodReadinessTimeout, mountpointPodReadinessCheckInterval, podMountSockPath)
+	if err != nil {
+		return fmt.Errorf("Failed to wait for Mountpoint Pod to be ready in given timeout for %q: %w", target, err)
+	}
+
+	klog.V(4).Infof("Mounting %s for %s", target, pod.Name)
+
+	fuseDeviceFD, err := pm.mountSyscallWithDefault(target, args)
+	if err != nil {
+		if fuseDeviceFD > 0 {
```

Comment:
Same comment as earlier, I think the check here doesn't make sense as file descriptor 0 is reserved for stdin